### PR TITLE
Try State Hook for Ranked Collective

### PR DESCRIPTION
**Pull request by dharjeezy**
Part of: paritytech/polkadot-sdk#239

Polkadot address: 12GyGD3QhT4i2JJpNzvMf96sxxBLWymz4RdGCxRH5Rj5agKW

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4982501

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4982647

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4982666

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4982665

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4982664

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5022447

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5022689

File: substrate/frame/ranked-collective/src/lib.rs [hunk 1]
```diff
@@ -763,6 +771,52 @@ pub mod pallet {
...
+				ensure!(
+				total_members as u32 >= member_index,
+				"Total count of `Members` should be greater than or equal to the number of `MemberIndex` of a particular `Rank` in `MemberCount`."
+				);
+
+				Ok(())
+			})?;
+
+			Ok(())
+		}
```
Comment (liamaharon) [inline substrate/frame/ranked-collective/src/lib.rs:801]: What about checking the rank in `MemberCount` matches the actual member rank?

File: substrate/frame/ranked-collective/src/lib.rs [hunk 2]
```diff
@@ -763,6 +771,52 @@ pub mod pallet {
...
+		/// ### Invariants of Index storage items
+		/// [`Member`] in storage of [`IdToIndex`] should be the same as [`Member`] in [`IndexToId`]
+		fn try_state_index() -> Result<(), sp_runtime::TryRuntimeError> {
+			IdToIndex::<T, I>::iter().try_for_each(
+				|(rank, who, member_index)| -> DispatchResult {
+					let who_from_index = IndexToId::<T, I>::get(rank, member_index).unwrap();
+					ensure!(
+				who == who_from_index,
+				"`Member` in storage of `IdToIndex` should be the same as `Member` in `IndexToId`."
+				);
```
Comment (liamaharon) [inline substrate/frame/ranked-collective/src/lib.rs:812]: Can we also check the rank here?

File: substrate/frame/ranked-collective/src/lib.rs [hunk 3]
```diff
@@ -763,6 +771,52 @@ pub mod pallet {
...
+		/// ### Invariants of Member storage items
+		///
+		/// Total number of [`Members`] in storage should be >= [`MemberIndex`] of a [`Rank`] in
+		///    [`MemberCount`].
+		#[cfg(any(feature = "try-runtime", test))]
+		fn try_state_members() -> Result<(), sp_runtime::TryRuntimeError> {
+			MemberCount::<T, I>::iter().try_for_each(|(_rank, member_index)| -> DispatchResult {
+				let total_members = Members::<T, I>::iter().count();
+				ensure!(
+				total_members as u32 >= member_index,
```
Comment (liamaharon) [inline substrate/frame/ranked-collective/src/lib.rs:793]: Is this >= or ==?

**Review by liamaharon — COMMENTED** [review]
Thanks for working on this!

Great start, but has a few more invariants to go.

File: substrate/frame/ranked-collective/src/lib.rs (hunk 3 above)
Comment (dharjeezy) [reply substrate/frame/ranked-collective/src/lib.rs:794]: This is >= because total_members in storage is always >= member_index

File: substrate/frame/ranked-collective/src/lib.rs (hunk 1 above)
Comment (dharjeezy) [reply substrate/frame/ranked-collective/src/lib.rs]: I have added an invariant for this

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5117866

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5117884

File: substrate/frame/ranked-collective/src/lib.rs [hunk 4]
```diff
@@ -838,6 +846,93 @@ pub mod pallet {
 			Members::<T, I>::remove(&who);
 			Ok(())
 		}
+
+		/// Ensure the correctness of the state of this pallet.
+		#[cfg(any(feature = "try-runtime", test))]
```
Comment (kianenigma) [inline substrate/frame/ranked-collective/src/lib.rs:851]: If you move this to a separate `impl` block it might look a bit neater, as we won't need to repeat the feature gate line.

**Review by kianenigma — APPROVED** [review]
I cannot be sure that all invariants are correctly captured here, but otherwise it is a good textbook example of hwat needs to be done. 

Can you check a possible invariant as with https://github.com/paritytech/polkadot-sdk/pull/3156#discussion_r1472679446? If there are a list of members being shared. 

Lastly, can you provide your DOT/KSM Address for a tip?

**Comment by dharjeezy** [comment]
> I cannot be sure that all invariants are correctly captured here, but otherwise it is a good textbook example of hwat needs to be done.
> 
> Can you check a possible invariant as with [#3156 (comment)](https://github.com/paritytech/polkadot-sdk/pull/3156#discussion_r1472679446)? If there are a list of members being shared.
>  
> Lastly, can you provide your DOT/KSM Address for a tip?

I have provided my Polkadot address in the PR description @kianenigma

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5121032

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5121174

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5121193

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5121192

File: substrate/frame/ranked-collective/src/lib.rs (hunk 3 above)
Comment (liamaharon) [reply substrate/frame/ranked-collective/src/lib.rs:870]: I think a better invariant to check here is that total_members equals the sum of all member_indexes?

File: substrate/frame/ranked-collective/src/lib.rs [hunk 5]
```diff
@@ -840,6 +848,91 @@ pub mod pallet {
...
+				|(rank, who, member_index)| -> DispatchResult {
+					let who_from_index = IndexToId::<T, I>::get(rank, member_index).unwrap();
+					ensure!(
+				who == who_from_index,
+				"`Member` in storage of `IdToIndex` should be the same as `Member` in `IndexToId`."
+				);
+
+					ensure!(
+						Self::is_rank_in_index_to_id_storage(rank.into()) == true,
+						"`Rank` in `IdToIndex` should be in `IndexToId`"
```
Comment (liamaharon) [inline substrate/frame/ranked-collective/src/lib.rs:903]: This doesn't make sense to me, maybe rephrase and/or add a comment

File: substrate/frame/ranked-collective/src/lib.rs [hunk 6]
```diff
@@ -840,6 +848,91 @@ pub mod pallet {
...
+			})?;
+
+			Members::<T, I>::iter().try_for_each(|(_, member_record)| -> DispatchResult {
+				ensure!(
+					Self::is_rank_in_member_count(member_record.rank.clone().into()) == true,
+					"`Rank` in Members should be in `MemberCount`"
+				);
+
+				Ok(())
+			})?;
```
Comment (liamaharon) [inline substrate/frame/ranked-collective/src/lib.rs:884]: This isn't a complete invariant for these two storage items. Please check that each MemberCount value is actually consistent with the number of Members at each rank.

File: substrate/frame/ranked-collective/src/lib.rs (hunk 2 above)
Comment (liamaharon) [reply substrate/frame/ranked-collective/src/lib.rs:899]: I mean please check that the rank of the member in IdToIndex is the same as the rank of the member in Members.

**Review by liamaharon — CHANGES_REQUESTED** [review]
This is close, but not capturing all invariants I would consider important to check. Remember, we want to ensure all the state across all storage items is consistent. If you have any questions about my comments feel free to ask.

File: substrate/frame/ranked-collective/src/lib.rs (hunk 5 above)
Comment (dharjeezy) [reply substrate/frame/ranked-collective/src/lib.rs:903]: so, for this @liamaharon https://github.com/paritytech/polkadot-sdk/pull/3007#discussion_r1479291547

The sum of all member_indexes will not equal total_members because, the indexes for a particular rank will keep increasing and not decrease. for example look at this test https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/ranked-collective/src/tests.rs#L456 the total_members are 3 and the sum of all the member_indexes is 9 because for every add or promote that is being done, therefore, the index of each rank will keep increasing if the member is in a particular rank or have been promoted. So, if a member was added at rank 0 and the promoted to rank 1, the index of rank 0 is 1 and the index of rank 1 is also 1 too, making the total index 2 for that particular member

File: substrate/frame/ranked-collective/src/lib.rs (hunk 6 above)
Comment (dharjeezy) [reply substrate/frame/ranked-collective/src/lib.rs:884]: So, each MemberCount value won't be consistent with the number of Members at each rank because of this statement here https://github.com/paritytech/polkadot-sdk/pull/3007#discussion_r1479457449

MemberCount holds indexes that keeps increasing and Member will always map an account to the rank that account belongs to at a particular time

**Comment by dharjeezy** [comment]
I have added an invariant to address your comment here
https://github.com/paritytech/polkadot-sdk/pull/3007#discussion_r1479298513
@liamaharon

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5128792

File: substrate/frame/ranked-collective/src/lib.rs (hunk 5 above)
Comment (dharjeezy) [reply substrate/frame/ranked-collective/src/lib.rs]: I have now rephrased this.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5129078

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5131643

File: substrate/frame/ranked-collective/src/lib.rs (hunk 3 above)
Comment (liamaharon) [reply substrate/frame/ranked-collective/src/lib.rs:877]: Ah ok, thanks for explaining. 

In this case, can we calculate what the index should be for each rank based on members?

e.g. if we have 1 member in rank 1 and one in rank 3, MemberCount should be

0: 2
1: 1
2: 1
?

Do I understand that right?

File: substrate/frame/ranked-collective/src/lib.rs (hunk 6 above)
Comment (liamaharon) [reply substrate/frame/ranked-collective/src/lib.rs:891]: Is it still possible to derive what MemberCount should be based on all the members? Or do I misunderstand

File: substrate/frame/ranked-collective/src/lib.rs (hunk 2 above)
Comment (liamaharon) [reply substrate/frame/ranked-collective/src/lib.rs:908]: Awesome, thanks

File: substrate/frame/ranked-collective/src/lib.rs (hunk 3 above)
Comment (dharjeezy) [reply substrate/frame/ranked-collective/src/lib.rs:877]: But then MemberCount actually already does this since it holds the storage of rank to index

File: substrate/frame/ranked-collective/src/lib.rs (hunk 6 above)
Comment (dharjeezy) [reply substrate/frame/ranked-collective/src/lib.rs]: Oh, yes. I have added the invariant for this. You can check again now.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5138909

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5138892

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5138910

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5138911

**Review by liamaharon — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5146634

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5146633

**Comment by liamaharon** [comment]
@dharjeezy can you take a look at the failing ci?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5153951
